Implement basic infrastructure to extract (primarily text) content from webpages

whsieh opened the pull request:
#### 569a8bd61f85e335544610a543eb86b808b09fa4
<pre>
Implement basic infrastructure to extract (primarily text) content from webpages
<a href="https://bugs.webkit.org/show_bug.cgi?id=268171">https://bugs.webkit.org/show_bug.cgi?id=268171</a>
<a href="https://rdar.apple.com/121132162">rdar://121132162</a>

Reviewed by Aditya Keerthi.

Add some infrastructure to WebKit, to extract visible text from web content for (eventual) donation
to system services. No change in behavior (yet).

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/text-extraction/TextExtraction.cpp: Added.
(WebCore::TextExtraction::collectText):

Add a utility function to collect text over the entire document, and then recursively walk the DOM
to collect any other elements that are interesting for the purposes of text extraction; note that
this skips subframes for the time being, and doesn&apos;t handle `RemoteFrame`. Support will be added in
subsequent patches.

(WebCore::TextExtraction::shouldIncludeChildren):
(WebCore::TextExtraction::rootViewBounds):
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::extractRecursive):
(WebCore::TextExtraction::extractItem):
* Source/WebCore/page/text-extraction/TextExtraction.h: Added.
* Source/WebCore/page/text-extraction/TextExtractionTypes.h: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestTextExtraction):

Add an unused `WebPageProxy` method and IPC endpoint for now that will be adopted in `WebViewImpl`
and `WKContentView` in subsequent patches to vend collected items to system services.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestTextExtraction):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/273598@main">https://commits.webkit.org/273598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71646313e49706c9c975cfcf8872bcc3e6b062ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38161 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38683 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32346 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37183 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17330 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11936 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36516 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31933 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11065 "Passed tests") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/36141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/32130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39931 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32672 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32452 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37031 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11244 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/9160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35113 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8191 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/11755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12075 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -0,0 +1,189 @@
+        return { { InteractiveItemData { !button->isDisabledFormControl() } } };
+
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(element)) {
+        if (input->isSubmitButton())
```
Should we use `input->isTextButton()` here?

`<input type=button>` won't be included for `isSubmitButton()`, and I don't see a good reason to exclude it, given that `HTMLButtonElement` is included. Unless, you also wanted to tie the inclusion of `HTMLButtonElement` to form submission?

whsieh replied:
Great catch — it should definitely be `isTextButton` instead of `isSubmitButton`.

pxlcoder commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -0,0 +1,189 @@
+static inline std::optional<ItemData> extractItemData(Node& node, const HashMap<Ref<Text>, String>& extractedText)
+{
+    if (!node.renderer())
+        return std::nullopt;
```
Should we exclude items with `visibility: hidden` or `opacity: 0`?

whsieh replied:
Hm... I think it makes sense to skip `visibility: hidden` elements, but for `opacity: 0` I think it's a bit difficult to say (since the text can still be selected and copied). I'll just bail on hidden for the time being.

Thanks for the review!

pxlcoder commented on Source/WebCore/page/text-extraction/TextExtraction.h:
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "TextExtractionTypes.h"
+#include <wtf/WeakPtr.h>
```
I don't think `#include <wtf/WeakPtr.h>` is necessary in the header?

whsieh replied:
Whoops, yes — this was left over from a prior iteration of this patch.

whsieh committed:
Implement basic infrastructure to extract (primarily text) content from webpages
https://bugs.webkit.org/show_bug.cgi?id=268171
rdar://121132162

Reviewed by Aditya Keerthi.

Add some infrastructure to WebKit, to extract visible text from web content for (eventual) donation
to system services. No change in behavior (yet).

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
* Source/WebCore/page/text-extraction/TextExtraction.cpp: Added.
(WebCore::TextExtraction::collectText):

Add a utility function to collect text over the entire document, and then recursively walk the DOM
to collect any other elements that are interesting for the purposes of text extraction; note that
this skips subframes for the time being, and doesn't handle `RemoteFrame`. Support will be added in
subsequent patches.

(WebCore::TextExtraction::shouldIncludeChildren):
(WebCore::TextExtraction::rootViewBounds):
(WebCore::TextExtraction::extractItemData):
(WebCore::TextExtraction::extractRecursive):
(WebCore::TextExtraction::extractItem):
* Source/WebCore/page/text-extraction/TextExtraction.h: Added.
* Source/WebCore/page/text-extraction/TextExtractionTypes.h: Added.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestTextExtraction):

Add an unused `WebPageProxy` method and IPC endpoint for now that will be adopted in `WebViewImpl`
and `WKContentView` in subsequent patches to vend collected items to system services.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::requestTextExtraction):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/273598@main
